donkeycar/parts/tub_v2.py

```diff
@@ -120,13 +145,55 @@ def run(self, *args):
         self.tub.write_record(record)
         return self.tub.manifest.current_index
 
-    def __iter__(self):
-        return self.tub.__iter__()
+    def run_threaded(self, *args):
+        assert len(self.tub.inputs) == len(args), \
+            f'Expected {len(self.tub.inputs)} inputs but received {len(args)}'
+        record = dict(zip(self.tub.inputs, args))
+        # push the (record, timestamp) tuple into queue
+        self.record_queue.put( (record, time.time()) )
+        self.mutex.acquire()
```

Comment:
No need for a mutex here. If you are worried about `current_index` gets being out of sync - that is always going to be the case as writes are happening on a dedicated thread. This won't help.

Reply:
One reader thread, one writer thread, when in doubt, use locks. I'm not worried `current_index` is "out of sync" nor I'm worry about `Queue` is not thread-safe. I'm following basic multi thread programming rule, to make the data store operation execute in sequence. Keep in mind Python can override attribute accessor quite easily, so there's no guarantee reading `current_index` is atomic nor it won't have interference with "writer" method like `write_record` without examine the implementation of `Manifest` thoroughly. So, without the knowledge of how `Manifest` works, it's better putting a lock around where we read anything from it, and around where we write anything to it. 

In addition, if you remove the lock, there won't be huge performance gain, but your code is more vulnerable to future changes.

Another side note, I've seen quite a lot places in this project with no proper thread protection, which makes the running wonky and sometimes unreliable. Understandable there are so many parts and implementations are from different contributor however it's better start to emphasise thread safety now.

Reply:
> One reader thread, one writer thread, when in doubt, use locks.

Use locks when thinking about contention. Not necessarily true otherwise. 
`current_index` being read from the manifest while writing is not going to be something that is going to be in sync even with locks.